File: packages/api/src/services/price-terms.ts
import { and, eq, gte } from 'drizzle-orm';
import { PublicKey } from '@solana/web3.js';
import {
  canonicalMint,
  isStableMint,
  lookupMint,
  NATIVE_SOL_PSEUDO_MINT,
  WRAPPED_SOL_MINT,
} from '@covantic/shared';
import type { Database } from '../config/database.js';
import { agentOutflowEvents } from '../db/schema.js';
import { FEED_IDS } from './oracle/price-sources/pyth-hermes.js';
import { CAP_HEADROOM_MULTIPLE } from './envelope-derivation.js';

/**
 * The asset an oracle-manipulation claim on a policy may be priced against,
 * as the oracle attests it and `create_policy` copies it into
 * `PolicyPriceTerms`.
 *
 * Derived, never chosen — by the holder or by the oracle at settlement. The
 * settlement instruction used to take the feed, the decimals and the quantity
 * from the oracle's own evidence and check them only against each other, so
 * a compromised oracle key could pick any genuine feed that had moved,
 * assert a position in it, and produce a loss bound that reached the
 * coverage for an asset the agent never touched. Fixing the terms at
 * purchase, from the agent's own record, is what closes that: the feed has
 * to be one the agent actually trades, and the quantity has to be in
 * proportion to what it actually moves.
 */
export interface AttestedPriceTermsInput {
  /** Pyth feed id, 32 bytes. All zero when the agent has no priced habit. */
  feedId: Uint8Array;
  subjectMint: PublicKey;
  subjectDecimals: number;
  /** Largest subject quantity a claim may assert, raw base units. */
  maxSubjectQuantity: bigint;
}

/** What `upsert_attestation` receives when there is nothing to price. */
export function emptyPriceTerms(): AttestedPriceTermsInput {
  return {
    feedId: new Uint8Array(32),
    subjectMint: PublicKey.default,
    subjectDecimals: 0,
    maxSubjectQuantity: 0n,
  };
}

export function isEmptyPriceTerms(terms: AttestedPriceTermsInput): boolean {
  return terms.feedId.every((b) => b === 0);
}

export interface SubjectObservation {
  mint: string;
  amountRaw: number;
  decimals: number;
}

/**
 * Pick the subject asset from what the agent has moved.
 *
 * Pure. The candidates are the non-stable mints the registry can price; the
 * one the agent moved most often is the habit worth insuring, and the largest
 * single movement in it — with the same headroom the envelope cap gets —
 * bounds what a claim may say was at stake. An agent with no priced habit
 * gets empty terms, and its oracle-manipulation claims go to a reviewer
 * rather than to an instruction that would have to take the asset on the
 * oracle's word.
 */
export function derivePriceTerms(observations: SubjectObservation[]): AttestedPriceTermsInput {
  const byMint = new Map<
    string,
    { count: number; largestRaw: number; decimals: number; feedKey: string }
  >();

  for (const o of observations) {
    if (!(o.amountRaw > 0)) continue;
    const mint = canonicalMint(o.mint);
    if (isStableMint(mint)) continue;
    const meta = lookupMint(mint);
    if (!meta || meta.kind !== 'priced' || !meta.feedKey || !FEED_IDS[meta.feedKey]) continue;
    // Registry decimals are authoritative; a recorded decimals that disagrees
    // is a scaling error waiting to happen, and the registry has been checked
    // against chain.
    const entry = byMint.get(mint) ?? {
      count: 0,
      largestRaw: 0,
      decimals: meta.decimals,
      feedKey: meta.feedKey,
    };
    entry.count += 1;
    entry.largestRaw = Math.max(entry.largestRaw, o.amountRaw);
    byMint.set(mint, entry);
  }

  let chosen: {
    mint: string;
    count: number;
    largestRaw: number;
    decimals: number;
    feedKey: string;
  } | null = null;
  for (const [mint, entry] of byMint) {
    if (
      chosen === null ||
      entry.count > chosen.count ||
      (entry.count === chosen.count && mint < chosen.mint)
    ) {
      chosen = { mint, ...entry };
    }
  }
  if (!chosen) return emptyPriceTerms();

  const feedHex = FEED_IDS[chosen.feedKey]!;
  // The registry folds wrapped SOL onto a pseudo-mint for valuation; the
  // account the program records has to be a real mint.
  const onChainMint = chosen.mint === NATIVE_SOL_PSEUDO_MINT ? WRAPPED_SOL_MINT : chosen.mint;
  return {
    feedId: Uint8Array.from(Buffer.from(feedHex, 'hex')),
    subjectMint: new PublicKey(onChainMint),
    subjectDecimals: chosen.decimals,
    maxSubjectQuantity: BigInt(Math.round(chosen.largestRaw * CAP_HEADROOM_MULTIPLE)),
  };
}

/** Load the agent's recent movements and derive its price terms. */
export async function loadPriceTerms(
  db: Database,
  agentAddress: string,
  now: Date,
  windowSeconds: number,
): Promise<AttestedPriceTermsInput> {
  const since = new Date(now.getTime() - windowSeconds * 1000);
  const rows = await db
    .select({
      mint: agentOutflowEvents.mint,
      amountRaw: agentOutflowEvents.amountRaw,
      decimals: agentOutflowEvents.decimals,
    })
    .from(agentOutflowEvents)
    .where(
      and(
        eq(agentOutflowEvents.agentAddress, agentAddress),
        gte(agentOutflowEvents.blockTime, since),
      ),
    );
  return derivePriceTerms(rows);
}

/** The shape the Anchor client serialises for `AttestedPriceTerms`. */
export function priceTermsToInstructionArg(terms: AttestedPriceTermsInput): {
  feedId: number[];
  subjectMint: PublicKey;
  subjectDecimals: number;
  maxSubjectQuantity: bigint;
} {
  return {
    feedId: Array.from(terms.feedId),
    subjectMint: terms.subjectMint,
    subjectDecimals: terms.subjectDecimals,
    maxSubjectQuantity: terms.maxSubjectQuantity,
  };
}

File: packages/anchor/programs/covantic/src/state/price_terms.rs
use anchor_lang::prelude::*;

use crate::constants::MAX_SUBJECT_DECIMALS;
use crate::errors::CovanticError;

/// The asset an oracle-manipulation claim on a policy may be priced against,
/// fixed at purchase and never changed.
///
/// `verify_and_payout_v2` proves a *price*: it reads a guardian-signed Pyth
/// update and refuses to pay more than the deviation from it can account for.
/// What it could not do before this account existed was say which price. The
/// feed, the quantity and the decimals all arrived in the oracle's evidence
/// and were checked only against each other, so a compromised oracle key could
/// pick any genuine feed that happened to have moved, assert a position in it,
/// and produce a loss bound that reached the coverage — for an asset the agent
/// never held.
///
/// The terms are attested by the oracle *before* the purchase, in the same
/// `RiskAttestation` that fixes the tier and prices the envelope, and copied
/// here by `create_policy`. A holder cannot choose them and the oracle cannot
/// change them afterwards. At settlement the evidence must name this feed,
/// this many decimals, and a quantity no larger than the bound recorded here.
///
/// A policy whose attestation carried no terms (`feed_id` all zero) is not
/// covered for oracle manipulation on the proof path: the instruction refuses
/// it, and the claim goes to a reviewer.
///
/// Kept in its own PDA rather than as new fields on `InsurancePolicy`, for the
/// reason `ClaimEvidenceRecord` gives: growing a live account type forces a
/// migration of every policy already on chain.
#[account]
pub struct PolicyPriceTerms {
    pub policy_id: u64,
    pub holder: Pubkey,

    /// Pyth feed the reference price must come from. All zero when the policy
    /// carries no oracle-manipulation terms.
    pub feed_id: [u8; 32],
    /// The asset that feed prices, so a reader can tie the feed to a mint.
    pub subject_mint: Pubkey,
    /// Decimals the evidence must scale the subject quantity with.
    pub subject_decimals: u8,
    /// Largest subject quantity a claim may assert, in the subject's base
    /// units. Bounds what the oracle can multiply a deviation by.
    pub max_subject_quantity: u64,

    pub bump: u8,
}

impl PolicyPriceTerms {
    pub const LEN: usize = 8   // discriminator
        + 8                    // policy_id
        + 32                   // holder
        + 32                   // feed_id
        + 32                   // subject_mint
        + 1                    // subject_decimals
        + 8                    // max_subject_quantity
        + 1; // bump

    /// Whether the policy carries terms a price claim can be settled against.
    pub fn is_priced(&self) -> bool {
        self.feed_id != [0u8; 32]
    }
}

/// The price terms as the oracle attests them.
///
/// Either wholly absent — every field zero, meaning the oracle priced no
/// oracle-manipulation cover for this agent — or wholly present. A partial
/// declaration is refused rather than interpreted.
#[derive(AnchorSerialize, AnchorDeserialize, Clone, Copy, Debug, Default, PartialEq, Eq)]
pub struct AttestedPriceTerms {
    pub feed_id: [u8; 32],
    pub subject_mint: Pubkey,
    pub subject_decimals: u8,
    pub max_subject_quantity: u64,
}

impl AttestedPriceTerms {
    pub fn is_empty(&self) -> bool {
        self.feed_id == [0u8; 32]
            && self.subject_mint == Pubkey::default()
            && self.subject_decimals == 0
            && self.max_subject_quantity == 0
    }

    pub fn validate(&self) -> Result<()> {
        if self.is_empty() {
            return Ok(());
        }
        require!(self.feed_id != [0u8; 32], CovanticError::InvalidPriceTerms);
        require!(
            self.subject_mint != Pubkey::default(),
            CovanticError::InvalidPriceTerms
        );
        require!(
            self.subject_decimals <= MAX_SUBJECT_DECIMALS,
            CovanticError::InvalidPriceTerms
        );
        require!(
            self.max_subject_quantity > 0,
            CovanticError::InvalidPriceTerms
        );
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_terms_are_valid_and_unpriced() {
        let terms = AttestedPriceTerms::default();
        assert!(terms.is_empty());
        assert!(terms.validate().is_ok());
    }

    #[test]
    fn a_partial_declaration_is_refused() {
        let terms = AttestedPriceTerms {
            feed_id: [1u8; 32],
            ..Default::default()
        };
        assert!(terms.validate().is_err());
    }

    #[test]
    fn a_full_declaration_is_accepted() {
        let terms = AttestedPriceTerms {
            feed_id: [1u8; 32],
            subject_mint: Pubkey::new_unique(),
            subject_decimals: 9,
            max_subject_quantity: 1_000,
        };
        assert!(!terms.is_empty());
        assert!(terms.validate().is_ok());
    }

    #[test]
    fn decimals_are_bounded() {
        let terms = AttestedPriceTerms {
            feed_id: [1u8; 32],
            subject_mint: Pubkey::new_unique(),
            subject_decimals: MAX_SUBJECT_DECIMALS + 1,
            max_subject_quantity: 1,
        };
        assert!(terms.validate().is_err());
    }
}
